### PR TITLE
cpio: add a cpio command

### PR DESCRIPTION
**Pull request by rminnich**
Currently this command only implements the t command, and only for
newc files. It's written in a way that it might be extended for
the other formats, but really, I doubt we care.

To add the other formats, you would add a switch for the type, and a
Writers map and Readers map which maps the string to the New functions
for those types. Not very hard, I doubt it's worth the effort. Before I
implement i and o commands, I want to get some tests in here. Or anyone who
wants to can contribute them.

I had looked at the sassoftware implementation of cpio, but it is
too complex, too limited, and quite unfinished. This is actually
much less code and already does more than that package does.

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

File: cmds/cpio/types.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+type RecReader interface {
+	RecRead() (*File, error)
+}
+
+type RecWriter interface {
+	RecWrite() (*File, error)
+}
+
+type File struct {
```
Comment (rjoleary) [inline cmds/cpio/types.go]: Place a comment here saying parsing will break if the fields are changed or reordered.

File: cmds/cpio/newc.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	f.Name = string(n[:f.NameSize-1])
+	if f.Name == "TRAILER!!!" {
+		debug("AT THE TRAILER!!!\n")
+		return nil, io.EOF
+	}
+
+	f.Data = &io.LimitedReader{t, int64(f.FileSize)}
+	seekTo = (int64(f.FileSize+3) / 4) * 4
```
Comment (rjoleary) [inline cmds/cpio/newc.go]: By casting to `int64`, you lose the ability to read from files >2gib. We probably won't see files that large anyways.

File: cmds/cpio/cpio.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+//
+//
+// Synopsis:
+//     cpio
+//
+// Description:
+//
+// Options:
+//     o: output an archive to stdout given a pattern
+//     i: output files from a stdin stream
```
Comment (rminnich) [inline cmds/cpio/cpio.go]: done

File: cmds/cpio/cpio.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+)
+
+var (
+	debug = func(string, ...interface{}) {}
+	d     = flag.Bool("d", false, "Debug prints")
```
Comment (rminnich) [inline cmds/cpio/cpio.go]: done, -v

File: cmds/cpio/cpio.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+)
+
+func usage() {
+	log.Fatalf("Usage: cpio")
```
Comment (rminnich) [inline cmds/cpio/cpio.go]: what do we do? flag.Usage won't cover all the eventual usage patterns?

File: cmds/cpio/func.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package main
+
+import "fmt"
+
+func (f *File) String() string {
+
+	return fmt.Sprintf("Ino 0x%x Mode 0%o UID %d GID %d Nlink %d Mtime %d FileSize %d Major %d Minor %d NameSize %d Name %s",
```
Comment (rminnich) [inline cmds/cpio/func.go]: done

**Comment by rminnich** [comment]
I've added one simple test but there will need to be more.

**Comment by rminnich** [comment]
btw I don't plan to add functionality for other cpio formats, but if we do, the reason for the interfaces will make more sense, as well as for a New for readers and writers: we might want to read a file format we don't want to write.

**Comment by rminnich** [comment]
This test result is nuts. The Go on travis and my go disagree on how to convert time ... the test does not fail on my machine.

File: cmds/cpio/cpio.go (hunk 5 above)
Comment (rjoleary) [inline cmds/cpio/cpio.go]: The `flag` package has default flags for `-h` and `-help` which call `flag.Usage` and exit, so overriding the function will print the same usage message using `-h` or with invalid arguments:

```
flag.Usage = func () {
    log.Fatalf("Usage: cpio")
}
```

Don't worry about fixing it now. If I have time later this week, I can go through each command and rewrite them this way.

**Pull request merged**